### PR TITLE
Refactor Analysis Server integration

### PR DESCRIPTION
**Pull request by lukechurch**
@devoncarew This takes a number of steps along the route we discussed.

It: 
- adds a completion benchmark
- adds the first end-to-end simple test of completion
- Refactors the Analysis Server into a wrapped class rather than using top level methods

There's still more to go, but this feels like a significant enough step in the right direction.

### CODE REVIEW

File: benchmark/bench.dart [hunk 1]
```diff
@@ -24,6 +26,10 @@ void main(List<String> args) {
     new AnalyzerBenchmark('hellohtml', sampleCodeWeb),
     new AnalyzerBenchmark('sunflower', _sunflower),
 
+    new AnalysisServerBenchmark('hello', sampleCode),
```
Comment (devoncarew) [inline benchmark/bench.dart]: What about calling this class `CompletionBenchmark`? The other class names (not part of this PR) should be updated at some point.

File: benchmark/bench.dart [hunk 2]
```diff
@@ -10,6 +10,8 @@ import 'package:services/src/analyzer.dart';
 import 'package:services/src/bench.dart';
 import 'package:services/src/common.dart';
 import 'package:services/src/compiler.dart';
+import 'package:services/src/analysis_server.dart';
```
Comment (devoncarew) [inline benchmark/bench.dart]: nit: sorting

File: benchmark/bench.dart (hunk 1 above)
Comment (devoncarew) [inline benchmark/bench.dart]: Also, we run each benchmark over three different types of sources. In the future, we should just drop this down to one moderately complex source file.

File: lib/src/analysis_server.dart [hunk 3]
```diff
@@ -1,182 +1,152 @@
...
-library services.completer_driver;
+/// A wrapper around an analysis server instance
+library services.analysis_server;
 
 import 'dart:io' as io;
 import 'dart:async';
 import 'dart:collection';
 import 'dart:convert';
 import 'package:logging/logging.dart';
+import 'dart:convert' show JSON;
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: nit: sorting - lexical and grouped by import type

File: lib/src/analysis_server.dart [hunk 4]
```diff
@@ -1,182 +1,152 @@
...
+      return new CompleteResponse(
+          response['replacementOffset'], response['replacementLength'],
+          results);
+    });
+  }
 
-  _onCompletionResults = new StreamController(sync: true);
-  completionResults = _onCompletionResults.stream.asBroadcastStream();
+  /// Cleanly shutdown the Analysis Server.
+  Future shutdown() async => serverConnection.sendServerShutdown();
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: Is it necessary to use the async keyword here?

File: lib/src/analysis_server.dart [hunk 5]
```diff
@@ -1,182 +1,152 @@
...
 
-  server.listenToOutput(dispatchNotification);
-  server.sendServerSetSubscriptions([ServerService.STATUS]);
+    return serverConnection.completionResults.handleError(
+        (error) => throw "Completion failed").first;
+  }
 
-  _logger.fine("Server Set Subscriptions completed");
+  /// Warm up the analysis server to be ready for use.
+  Future warmup([bool useHtml = false]) =>
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: possibly prefer a named optional param here

File: lib/src/analysis_server.dart [hunk 6]
```diff
@@ -465,3 +471,34 @@ class Server {
   }
 }
 
+class CompleteResponse {
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: `CompleteResponse` or `CompletionResponse`?

Comment (devoncarew) [inline lib/src/analysis_server.dart]: Also, I think all the rpc response classes should go into the common_server.dart file.

Comment (devoncarew) [inline lib/src/analysis_server.dart]: Ah, it was CompleteResonse before.

File: test/analysis_server_test.dart [hunk 7]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library services.analyzer_server_test;
+
+import 'package:services/src/analysis_server.dart';
+import 'package:grinder/grinder.dart' as grinder;
```
Comment (devoncarew) [inline test/analysis_server_test.dart]: You could use the `cli_utils` package here instead.

File: test/analysis_server_test.dart [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  group('analysis_server', () {
+    setUp(() {
+      analysisServer = new AnalysisServerWrapper(sdkPath);
+    });
+
+    tearDown(() {
+      analysisServer.shutdown();
+    });
+
+    test('simple', () {
```
Comment (devoncarew) [inline test/analysis_server_test.dart]: woot!

**Comment by devoncarew** [comment]
lgtm! I saw a build failure - I re-started the bot as it wasn't clear what happened. I _think_ all the tests ran, but then it hung when starting to run the benchmarks.

File: test/common_server_test.dart [hunk 9]
```diff
@@ -84,11 +84,12 @@ void defineTests() {
           contains('failed with errors: [error, line 2] Expected'));
     });
 
-    /* XXX: Disabled below tests as they are causing the dart vm to crash.
     test('complete', () async {
       var json = {'source': 'void main() {print("foo");}', 'offset': 1};
       var response = await _sendPostRequest('dartservices/v1/complete', json);
-      expect(response.status, 501);
+      expect(response.status, 200);
```
Comment (devoncarew) [inline test/common_server_test.dart]: You could try commenting these three tests back out. Perhaps the AS is not being terminated, and the VM is just never exiting (waiting on an open socket)?

File: benchmark/bench.dart (hunk 1 above)
Comment (lukechurch) [inline benchmark/bench.dart]: sg. Will address in follow on PR.

File: benchmark/bench.dart (hunk 2 above)
Comment (lukechurch) [inline benchmark/bench.dart]: done

File: lib/src/analysis_server.dart (hunk 3 above)
Comment (lukechurch) [inline lib/src/analysis_server.dart]: Hmmm, there really shouldn't be two separate imports for this. Fixed.

File: lib/src/analysis_server.dart (hunk 4 above)
Comment (lukechurch) [inline lib/src/analysis_server.dart]: Nope, legacy. Thanks, removed.

File: lib/src/analysis_server.dart (hunk 5 above)
Comment (lukechurch) [inline lib/src/analysis_server.dart]: This is consistent with the other APIs, however, I agree. I'll adjust them all in one go.

File: test/analysis_server_test.dart (hunk 7 above)
Comment (lukechurch) [inline test/analysis_server_test.dart]: done. I've adjusted the other test targets that used grinder as well.

File: test/common_server_test.dart (hunk 9 above)
Comment (lukechurch) [inline test/common_server_test.dart]: Yeah, I think you're probably right. Done.

Comment (lukechurch) [inline test/common_server_test.dart]: There was also a problem with the benchmarks' method being called 'teardown' not 'tearDown'.
